Planner integration

orduno opened the pull request:
- Created the DijkstraPlanner based on the core algorithms from Navfn ROS1 nav package.
- Updated the SimpleNavigator to use the new DijkstraPlanner
- Created an initial WorldModel node and Costmap mockup to test the new planner.
- Fixed bug with TaskClient.
- Added a few new messages and services related to Costmap. Made a few updates to the Path messages.

mjeronimo commented on src/costmap/CMakeLists.txt:
```diff
@@ -0,0 +1,38 @@
+endif()
+
+find_package(ament_cmake REQUIRED)
+# find_package(rclcpp REQUIRED)
```
Please remove commented-out sections to keep things tidy.

orduno replied:
OK

mjeronimo commented on src/costmap/CMakeLists.txt:
```diff
@@ -0,0 +1,38 @@
+
+include_directories(include)
+
+add_library(costmap STATIC
```
We may want a general utility library that can hold Costmap, etc., instead of proliferating directories for each library.

orduno replied:
Agree, made a note to discuss this on our next meeting.

orduno replied:
As discussed, i moved it to a util library

mjeronimo commented on src/costmap/include/costmap/CostValues.h:
```diff
@@ -0,0 +1,47 @@
+#ifndef WORLD_MODEL__COST_VALUES_H_
+#define WORLD_MODEL__COST_VALUES_H_
+/** Provides a mapping for often used cost values */
+namespace costmap
```
Please run ament_cpplint and ament_uncrustify on all code and address issues. It may not be practical yet for old code (Navfn), but should do as much as possible.

orduno replied:
Will do.

mjeronimo commented on src/costmap/include/costmap/Costmap.hpp:
```diff
@@ -0,0 +1,38 @@
+#ifndef COSTMAP__COSTMAP_HPP_
+#define COSTMAP__COSTMAP_HPP_
+
+#include <vector>
```
I don't see that vector is required in the interface. Should remove if not. Can include this in the implementation file, if necessary.

mjeronimo commented on src/costmap/package.xml:
```diff
@@ -0,0 +1,21 @@
+  <version>0.1.0</version>
+  <description>TODO</description>
+  <author email="oregon.robotics.team@intel.com">Oregon Robotics Team</author>
+   <author email="carlos.a.orduno@intel.com">Carlos Orduno</author>
```
Indentation is off. Also should discuss (again) how we're handing author and maintainer fields.

orduno replied:
Added to list of discussion topics.

mjeronimo commented on src/costmap/src/Costmap.cpp:
```diff
@@ -0,0 +1,161 @@
+  RCLCPP_INFO(node_->get_logger(), "Costmap::getCostmap");
+
+  // TODO(orduno): faking out a costmap for now
+  (void)specifications;
```
Instead of casting to void, you can also comment out the name ("specifications") in the function interface (line 27). I prefer commenting out the name as it results in fewer lines of code.

orduno replied:
Agree

mjeronimo commented on src/costmap/src/Costmap.cpp:
```diff
@@ -0,0 +1,161 @@
+
+  // Some arbitrary numbers
+  costmap.info.resolution = 1;  // m/cell
+  costmap.info.width = 10;  // cells
```
Can we change the name of the field from "width" to "cellWidth" to make it clearer (and not require the comment)?

orduno replied:
Agree on the principle. In this case, a better name is number_cells_x_direction, same for height. However I took the variable name as a carry over from Occupancy Grid, and as we see on other ROS message definitions, the preference is to use something short. Added to the list of things to discuss.

mjeronimo commented on src/costmap/src/Costmap.cpp:
```diff
@@ -0,0 +1,161 @@
+Costmap::getTestData(const int width, const int height, vector<uint8_t>& data)
+{
+  // TODO(orduno): fixed size for now
+  (void)width;
```
Comment out names on line 68?

orduno replied:
Fixed

mjeronimo commented on src/costmap/src/Costmap.cpp:
```diff
@@ -0,0 +1,161 @@
+
+  // TODO(orduno): instead of hardcoding, define a function
+
+  const uint8_t n = 255; // no information
```
Each of these special values should have a symbol. It could be from an enumeration (with explicit values) or a macro. That way, the code wouldn't need comments and would be clearer.

orduno replied:
Changed to enum.

orduno replied:
After further discussion, it was changed to static const within Costmap class

mjeronimo commented on src/costmap/src/Costmap.cpp:
```diff
@@ -0,0 +1,161 @@
+#include "costmap/Costmap.hpp"
+#include <tf2/LinearMath/Quaternion.h>
+
+using std::vector;
```
We should discuss our team conventions on whether we're opening namespaces (std) or not. We should be consistent as a team (and as individuals).

orduno replied:
I'd be in favor of opening with some rules: not on header files, only introducing the name being used. Added to the discussion list.

mjeronimo commented on src/costmap/src/Costmap.cpp:
```diff
@@ -0,0 +1,161 @@
+
+  // TODO(orduno): instead of hardcoding, define a function
+
+  const uint8_t n = 255; // no information
```
Should have a symbol for 255.

mjeronimo commented on src/nav2_msgs/msg/Costmap.msg:
```diff
@@ -0,0 +1,9 @@
+std_msgs/Header header
+
+# MetaData for the map
+CostmapMetaData info
```
Perhaps the field should be called "metadata" instead of "info."

orduno replied:
This is related to a previous comment about ROS message definitions and carryover from ROS1. Added to the discussion list.

orduno replied:
Fixed

mjeronimo commented on src/nav2_msgs/msg/Costmap.msg:
```diff
@@ -0,0 +1,9 @@
+CostmapMetaData info
+
+# The cost data, in row-major order, starting with (0,0).
+uint8[] data
```
Seems like there needs to be newline here.

orduno replied:
Added

mjeronimo commented on src/nav2_msgs/msg/CostmapMetaData.msg:
```diff
@@ -0,0 +1,23 @@
+
+# The origin of the costmap [m, m, rad].
+# This is the real-world pose of the cell (0,0) in the map.
+geometry_msgs/Pose origin
```
newline

orduno replied:
Added

mjeronimo commented on src/nav2_msgs/msg/PathEndPoints.msg:
```diff
@@ -1,3 +1,11 @@
-geometry_msgs/PoseWithCovariance goal
+
+# The start pose for the plan
+geometry_msgs/PoseStamped start
```
Why do we need a stamped version of the pose? We have the timestamp in the header already. Same for the goal.

orduno replied:
Currently not used. In future we might want to add time constraints to start and reach goal.

orduno replied:
Fixed, Mike is right, posestamp is not needed, and we don't want to overload the term. If we need to add time constraints we'll have to add a new field.

mjeronimo commented on src/nav2_msgs/srv/GetCostmap.srv:
```diff
@@ -0,0 +1,6 @@
+# Specifications for the requested costmap
+nav2_msgs/Costmap map
+---
+nav2_msgs/Costmap map
```
newline.

orduno replied:
Fixed. Also realized that the request should be nav2_msgs/CostmapMetaData -- no need for the data field.

mjeronimo commented on src/navigation/src/SimpleNavigator.cpp:
```diff
@@ -12,17 +12,23 @@
 #include <memory>
 #include <exception>
 #include <chrono>
+#include <iostream>
```
I don't see that iostream is required. Still using it? Should remove any #includes that aren't required.

orduno replied:
Removed

mjeronimo commented on src/planning/include/planning/AStarPlanner.hpp:
```diff
@@ -28,4 +28,4 @@ class AStarPlanner : public ComputePathToPoseTaskServer
 };
 
-#endif  // PLANNING__ASTARPLANNER_HPP_
+#endif  // PLANNING__ASTARPLANNER_HPP_
```
newline

mjeronimo commented on src/planning/include/planning/DijkstraPlanner.hpp:
```diff
@@ -0,0 +1,109 @@
+  bool computePotential(const geometry_msgs::msg::Point& world_point);
+
+  // Compute a plan to a goal from a potential - must call computePotential first
+  bool getPlanFromPotential(const geometry_msgs::msg::PoseStamped& goal, std::vector<geometry_msgs::msg::PoseStamped>& plan);
```
Please run (all files) through ament_cpplint and ament_uncrustify. Several lines are too long.

mjeronimo commented on src/planning/include/planning/Navfn.hpp:
```diff
@@ -0,0 +1,267 @@
+   * @class NavFn
+   * @brief Navigation function class. Holds buffers for costmap, navfn map. Maps are pixel-based. Origin is upper left, x is right, y is down. 
+   */
+  class NavFn
```
We should refactor this class. It's not clear to me what it is. We can probably extract out the various ideas into different (reusable) classes. Doesn't have to be done immediately though.

mjeronimo commented on src/planning/include/planning/Planner.hpp:
```diff
@@ -0,0 +1,32 @@
+// Copyright (c) 2018 Intel Corporation
```
This file is obsolete and can be removed.

orduno replied:
Removed.

mjeronimo commented on src/planning/package.xml:
```diff
@@ -5,7 +5,9 @@
   <version>0.1.0</version>
   <description>TODO</description>
   <author email="oregon.robotics.team@intel.com">Oregon Robotics Team</author>
+   <author email="carlos.a.orduno@intel.com">Carlos Orduno</author>
```
indentation off.

orduno replied:
Fixed

mjeronimo commented on src/planning/src/AStarPlanner.cpp:
```diff
@@ -55,4 +55,4 @@ AStarPlanner::executeAsync(const ComputePathToPoseCommand::SharedPtr command)
 
   return TaskStatus::SUCCEEDED;
-}
+}
```
newline

orduno replied:
Added, not sure how i removed those...

mjeronimo commented on src/planning/src/DijkstraPlanner.cpp:
```diff
@@ -0,0 +1,436 @@
+  }
+  catch (...) {
+    RCLCPP_ERROR(this->get_logger(), "DijkstraPlanner::makePlan: failed to obtain costmap from server");
+    throw;
```
I recommend throwing a specific kind of exception. See the standard exception types.

orduno replied:
The intention here is to re-throw the current exception.

mjeronimo commented on src/planning/src/DijkstraPlanner.cpp:
```diff
@@ -0,0 +1,436 @@
+  planner_ = std::make_shared<NavFn>(costmap_.info.width, costmap_.info.height);
+
+  // Plan publisher for visualization purposes
+  plan_publisher_ = this->create_publisher<nav2_msgs::msg::Path>("plan", 1);
```
Did you intend to include the second parameter (qos_history_depth)? Or, is this a left-over from porting?

orduno replied:
left-over from porting

mjeronimo commented on src/planning/src/Logger.h:
```diff
@@ -0,0 +1,17 @@
+#pragma once
```
Should port the code to not require this header (and remove it).

orduno replied:
Yes, also needs to be moved to a util folder

mjeronimo commented on src/planning/src/main.cpp:
```diff
@@ -15,11 +15,13 @@
 {
   rclcpp::init(argc, argv);
-  rclcpp::spin(std::make_shared<AStarPlanner>("AStarPlanner"));
+  // rclcpp::spin(std::make_shared<AStarPlanner>("AStarPlanner"));
```
Remove commented-out code.

orduno replied:
done

mjeronimo commented on src/task/include/task/TaskClient.hpp:
```diff
@@ -60,7 +61,7 @@ class TaskClient
 
   // The client can wait for a result with a timeout
-  TaskStatus waitForResult(typename ResultMsg::SharedPtr result, unsigned int milliseconds)
+  TaskStatus waitForResult(typename ResultMsg::SharedPtr& result, unsigned int milliseconds)
```
From the discussion with Carl in our team meeting, we want to use the first version (without the reference).

orduno replied:
There are some issues with the function without the reference. I found [this](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rr-sharedptrparam).

crdelsey replied:
Looking at Carlos' link and thinking some more about it, it seems like the normal usage is to pass as a const &.
Exceptions:
1. You want to share ownership with the callee. Then pass by value. This would apply when the callee is going to hang on to the shared_ptr for some time beyond the length of that one function call.
2. You are passing the shared_ptr as an output parameter. Then pass by reference.

This particular instance looks like exception 2.

mjeronimo commented on src/world_model/package.xml:
```diff
@@ -0,0 +1,22 @@
+  <version>0.1.0</version>
+  <description>TODO</description>
+  <author email="oregon.robotics.team@intel.com">Oregon Robotics Team</author>
+   <author email="carlos.a.orduno@intel.com">Carlos Orduno</author>
```
indentation

orduno replied:
fixed

mjeronimo commented on src/world_model/src/WorldModel.cpp:
```diff
@@ -0,0 +1,39 @@
+{
+  costmap_ = std::make_unique<Costmap>(this);
+
+  // Create a (lambda) callback function for when a costmap service is received.
```
Why a lambda function here versus a class method? Is there an advantage?

orduno replied:
I'm having problems defining a ROS2 service that takes in a class method for the callback.

mjeronimo commented on src/world_model/src/main.cpp:
```diff
@@ -0,0 +1,26 @@
+  rclcpp::shutdown();
+
+  return 0;
+}
```
newline

orduno replied:
fixed

crdelsey commented on src/costmap/CMakeLists.txt:
```diff
@@ -0,0 +1,35 @@
+  costmap
+  ARCHIVE DESTINATION lib
+  LIBRARY DESTINATION lib)
+
```
You should add these lines here so other packages can find this one.
`ament_export_include_directories(include)`
`ament_export_libraries(costmap)`

orduno commented on src/task/include/task/TaskClient.hpp:
```diff
@@ -48,6 +48,7 @@ class TaskClient
   void executeAsync(const typename CommandMsg::SharedPtr msg)
   {
     taskSucceeded_ = false;
+    receivedNewMsg_ = false;
```
changed it to receivedResult_

orduno commented on src/task/include/task/TaskClient.hpp:
```diff
@@ -60,7 +61,7 @@ class TaskClient
-  TaskStatus waitForResult(typename ResultMsg::SharedPtr result, unsigned int milliseconds)
+  TaskStatus waitForResult(typename ResultMsg::SharedPtr& result, unsigned int milliseconds)
   {
     std::mutex m;
```
@mjeronimo Can you look at @crdelsey 's question

orduno commented on src/task/include/task/TaskClient.hpp:
```diff
@@ -115,6 +120,8 @@ class TaskClient
 
   // A convenience variable for whether the task succeeded or not
   bool taskSucceeded_;
```
@mjeronimo Can you confirm @crdelsey 's comment?

orduno commented:
- Created the DijkstraPlanner based on the core algorithms from Navfn ROS1 nav package.
- Updated the SimpleNavigator to use the new DijkstraPlanner
- Created an initial WorldModel node and Costmap mockup to test the new planner.
- Fixed bug with TaskClient.
- Added a few new messages and services related to Costmap. Made a few updates to the Path messages.
